### PR TITLE
Externalize the extension configuration templates

### LINKED ISSUE #91: Externalize the extension configuration templates
## Feature Request

Currently the extension configuration templates like [nginx/template.go](https://github.com/ehazlett/interlock/blob/master/ext/nginx/template.go) and [haproxy/template.go](https://github.com/ehazlett/interlock/blob/master/ext/haproxy/template.go) are compiled right into Interlock. That means you're locked into the how that configuration is done.

The extension configuration templates should be externalized as simple text files that can make use of Interlock variables (e.g. `{{ $host.Name }}`). You could then include the configuration templates in your `Dockerfile` to be consumed by Interlock.

### PR DESCRIPTION
**Pull request by everett-toews**
Closes #91

### CODE REVIEW

File: docs/examples/haproxy/config.toml [hunk 1]
```diff
@@ -3,8 +3,9 @@ dockerURL = "unix:///var/run/docker.sock"
 
 [[extensions]]
 name = "haproxy"
-configPath = "/usr/local/etc/haproxy.cfg"
-pidPath = "/etc/haproxy.pid"
+configPath = "/usr/local/etc/haproxy/haproxy.cfg"
```
Comment (ehazlett) [inline docs/examples/haproxy/config.toml]: Is this the actual path of the config?  I thought it was at `/usr/local/etc/haproxy.cfg`.

Comment (everett-toews) [inline docs/examples/haproxy/config.toml]: Ya. [haproxy.org isn't too helpful](https://www.google.com/webhp?sourceid=chrome-instant&ion=1&espv=2&ie=UTF-8#q=site:www.haproxy.org+%22haproxy.cfg%22). I think it mostly comes down where distros package it, seems like [Ubuntu](https://www.google.com/webhp?sourceid=chrome-instant&ion=1&espv=2&ie=UTF-8#q=ubuntu%20haproxy.cfg) and [CentOS](https://www.google.com/webhp?sourceid=chrome-instant&ion=1&espv=2&ie=UTF-8#q=centos%20haproxy.cfg) favour `/etc/haproxy/haproxy.cfg`.

That's always been my personal preference too (e.g. /etc/appname/appname.conf, /etc/appname/appname.yml, etc.). But honestly I don't feel strongly about it here. I'm fine changing it back to `/usr/local/etc/haproxy.cfg` if that's your preference.

Comment (ehazlett) [inline docs/examples/haproxy/config.toml]: Since we recommend the official Docker hub HAProxy image and we run in
containers we should use what that image uses. I'll take a look to confirm.
Thanks!

**Comment by aquavitae** [comment]
FWIW, The official image uses /usr/local/etc/haproxy/haproxy.cfg

**Comment by ehazlett** [comment]
@aquavitae thanks -- @everett-toews let's keep it with what the official image is.. thanks!!

File: cmd/interlock/run.go [hunk 2]
```diff
@@ -163,18 +163,27 @@ func runAction(c *cli.Context) {
...
 	} else {
 		configPath := c.String("config")
+
 		if configPath != "" {
 			log.Debugf("loading config from: %s", configPath)
 
 			d, err := ioutil.ReadFile(configPath)
 			switch {
 			case os.IsNotExist(err):
+				log.Infof("Missing Interlock configuration: file=%s", configPath)
```
Comment (ehazlett) [inline cmd/interlock/run.go]: These should probably be `Error` level.

Comment (ehazlett) [inline cmd/interlock/run.go]: You might also want to collapse this into a multiline string with a single `Error`.

File: cmd/interlock/run.go [hunk 3]
```diff
@@ -163,18 +163,27 @@ func runAction(c *cli.Context) {
...
+				log.Info("Use the run --config option to set a custom location for the configuration file")
+				log.Info("Examples of an Interlock configuration file: url=https://github.com/ehazlett/interlock/tree/master/docs/examples")
 				log.Fatalf("config not found: %s", configPath)
 			case err == nil:
 				data = string(d)
 			default:
 				log.Fatal(err)
 			}
+		} else {
```
Comment (ehazlett) [inline cmd/interlock/run.go]: Can we DRY this out?

File: config.toml.example [hunk 4]
```diff
@@ -1,40 +0,0 @@
-# sample config for interlock
```
Comment (ehazlett) [inline config.toml.example]: Why did you remove the example?

File: config/utils_test.go [hunk 5]
```diff
@@ -28,7 +28,7 @@ func TestParseConfig(t *testing.T) {
 
 func TestSetConfigDefaults(t *testing.T) {
 	cfg := &ExtensionConfig{
-		Name: "test",
+		Name: "haproxy",
```
Comment (ehazlett) [inline config/utils_test.go]: Can we add more tests here to check the different config default setting on the backend?

File: ext/lb/haproxy/haproxy.go [hunk 6]
```diff
@@ -43,7 +44,17 @@ func (p *HAProxyLoadBalancer) ConfigPath() string {
 }
 
 func (p *HAProxyLoadBalancer) Template() string {
-	return haproxyConfTemplate
+	d, err := ioutil.ReadFile(p.cfg.TemplatePath)
+
+	if err == nil {
+		return string(d)
+	} else {
+		log().Infof("Missing haproxy configuration template: file=%s", p.cfg.TemplatePath)
```
Comment (ehazlett) [inline ext/lb/haproxy/haproxy.go]: We have this spread throughout.  We should not `log.Fatal` in the extensions.  This should be updated to return `(string, error)` and then bubble the error up.

File: ext/lb/nginx/nginx.go [hunk 7]
```diff
@@ -44,11 +45,17 @@ func (p *NginxLoadBalancer) HandleEvent(event *dockerclient.Event) error {
...
 
 func (p *NginxLoadBalancer) Template() string {
-	if p.cfg.NginxPlusEnabled {
-		return nginxPlusConfTemplate
+	d, err := ioutil.ReadFile(p.cfg.TemplatePath)
+
+	if err == nil {
+		return string(d)
+	} else {
+		log().Infof("Missing nginx configuration template: file=%s", p.cfg.TemplatePath)
```
Comment (ehazlett) [inline ext/lb/nginx/nginx.go]: We have this spread throughout. We should not log.Fatal in the extensions. This should be updated to return `(string, error)` and then bubble the error up.

File: cmd/interlock/run.go (hunk 3 above)
Comment (everett-toews) [inline cmd/interlock/run.go]: I brought a towel. This got me thinking it would be very helpful to have a sane default for the location of config.toml so I updated `cli.Command`.

File: cmd/interlock/run.go (hunk 2 above)
Comment (everett-toews) [inline cmd/interlock/run.go]: Changed to `Error` level.

I tried a multiline string like this.

```
func missingConfig(configPath string)  {
	log.Errorf(`Missing Interlock configuration: file=%s
			    Use the run --config option to set a custom location for the configuration file
                Examples of an Interlock configuration file: url=https://github.com/ehazlett/interlock/tree/master/docs/examples`, configPath)
	log.Fatalf("config not found: file=%s", configPath)
}
```

But that produced the output.

```
# cmd/interlock/interlock -D run --config blah
INFO[0000] interlock 1.2.0-dev (HEAD)
ERRO[0000] Missing Interlock configuration: file=blah
	            Use the run --config option to set a custom location for the configuration file
	            Examples of an Interlock configuration file: url=https://github.com/ehazlett/interlock/tree/master/docs/examples
FATA[0000] config not found: file=blah
```

Not only does it have the funky indentation but I think the missing `ERRO` at the beginning of the lines could confuse log parsers setup to work with Interlock.

File: config.toml.example (hunk 4 above)
Comment (everett-toews) [inline config.toml.example]: I moved it to `docs/examples/**/config.toml`

I felt it's better to have this example alongside all of the other examples so users only have to go to one place to find examples.

Also, having only one example in the root is a bit confusing because it has to cover multiple backends. For example, the nginx extension below includes a number of config options like `ConnectTimeout` and `ServerTimeout` that are specific to haproxy.

File: config/utils_test.go (hunk 5 above)
Comment (everett-toews) [inline config/utils_test.go]: Done.

File: docs/examples/haproxy/config.toml (hunk 1 above)
Comment (everett-toews) [inline docs/examples/haproxy/config.toml]: Left it as `/usr/local/etc/haproxy/haproxy.cfg` as per the [haproxy Dockerfile](https://github.com/docker-library/haproxy/blob/master/1.6/Dockerfile#L32)

File: ext/lb/haproxy/haproxy.go (hunk 6 above)
Comment (everett-toews) [inline ext/lb/haproxy/haproxy.go]: Good idea. This got me thinking it doesn't fail fast enough too. 

I moved the check for this up to `lb.NewLoadBalancer` so it's DRY and fails faster.

**Comment by ehazlett** [comment]
This needs a final rebase and then I think we are good.  FYI, there was just a PR merged that affected the templates.

**Comment by everett-toews** [comment]
Okay. I'll merge the changes from master and make sure everything is okay.

**Comment by everett-toews** [comment]
Looks good. Shall I squash this down?

File: cmd/interlock/run.go (hunk 3 above)
Comment (ehazlett) [inline cmd/interlock/run.go]: Ok I will need to update the env var parsing then as it checks if a config is passed.  Since you specific a default this will break.

File: cmd/interlock/run.go [hunk 8]
```diff
@@ -35,7 +35,7 @@ var cmdRun = cli.Command{
 		cli.StringFlag{
 			Name:  "config, c",
 			Usage: "path to config file",
-			Value: "",
+			Value: "/etc/interlock/config.toml",
```
Comment (ehazlett) [inline cmd/interlock/run.go]: As mentioned in another line, this will break the env var parsing.  Since we don't include a default I think we should remove this an error if no config is found.

Comment (ehazlett) [inline cmd/interlock/run.go]: This also means that the Docker command needs to be explicit which I think is also better (explicit vs. implicit) for configuration.

**Comment by ehazlett** [comment]
Minor nit with the config path.  Otherwise looks good.

**Comment by everett-toews** [comment]
Not a minor nit at all! Sorry I missed the consequences of doing a default config location in your original comment!

I updated the code and docs to try to make it clear what config styles are possible.

**Comment by everett-toews** [comment]
@ehazlett Is there anything else that needs to happen here before merging?

**Comment by ehazlett** [comment]
lgtm thanks!

**Comment by ehazlett** [comment]
I had to revert this.  It broke every existing installation.  I will be carrying this in another PR.  Thx.

**Comment by everett-toews** [comment]
Ouch but I suppose that was to be expected. No one is going to have an external configuration template. I had viewed this as a change for a 2.0 image version because it is a breaking change. I should have called that out explicitly.

I see there's a `master` tag for the image now. That tag was created after this PR originally started and I hadn't noticed it's creation. When you say "It broke every existing installation", do you mean it broke every existing installation running on `ehazlett/interlock:master`?

 I'm definitely in agreement on reverting it until there's a plan to roll it out to users.

**Comment by ehazlett** [comment]
Yes I hadn't released but tested on my infra and it broke. I didn't realize
you had removed the template completely. That was my fault. I'm planning to
carry this and add back the internal generation to not break backwards
compat but also add the ability to specify external templates. No worries!
We'll get it for the upcoming release :)

**Comment by everett-toews** [comment]
I'm glad you caught it before it rolled out to all users! I would have felt pretty bad about breaking that many users. 😰

**Comment by ehazlett** [comment]
Ha no worries.  I also have on the list to get an integration test rig up
as well.  That will help a bunch :)

**Comment by everett-toews** [comment]
You're certainly welcome to use [Carina](https://getcarina.com/) to create test clusters.

**Comment by everett-toews** [comment]
Thinking about it...adding back the internal generation to not break backwards is pretty simple. I could knock that out pretty quickly if you like.

**Comment by ehazlett** [comment]
I think that would be best. Allow external to be specified but not
required.
